### PR TITLE
Add an attribute to rename definitions in the backend

### PR DESCRIPTION
**Pull request by EschericHya**
Fixes https://github.com/AeneasVerif/charon/issues/207

Still need to write some tests I guess

### CODE REVIEW

File: charon/src/translate/translate_ctx.rs [hunk 1]
```diff
@@ -451,12 +451,39 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
             .any(|attr| attr == "charon::opaque" || attr == "aeneas::opaque");
+        let rename = {
+            let pos = attributes.iter().position(|str| {
+                str.starts_with("charon::rename(") || str.starts_with("aeneas::rename(")
+            });
+            if pos.is_none() {
+                None
+            } else {
+                let str = match attributes
+                    .get(pos.expect("Should not be None, it was checked right before"))
```
Comment (sonmarcho) [inline charon/src/translate/translate_ctx.rs:462]: You should simply use `Option::unwrap`

**Review by sonmarcho — CHANGES_REQUESTED** [review]

File: charon/src/translate/translate_ctx.rs [hunk 2]
```diff
@@ -452,28 +452,43 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
             .iter()
             .any(|attr| attr == "charon::opaque" || attr == "aeneas::opaque");
         let rename = {
-            let pos = attributes.iter().position(|str| {
+            let str = attributes.iter().find(|str| {
                 str.starts_with("charon::rename(") || str.starts_with("aeneas::rename(")
             });
-            if pos.is_none() {
+            if str.is_none() {
```
Comment (sonmarcho) [inline charon/src/translate/translate_ctx.rs:458]: You can use the `if let ...` notation (you use it below actuallly) instead of `str.is_none()` then `str.unwrap()`.

File: charon/src/translate/translate_ctx.rs [hunk 3]
```diff
@@ -452,28 +452,43 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                    if !str.is_empty() {
+                        let first_char_alphabetic = str
+                            .chars()
+                            .nth(0)
+                            .expect("Attribute `rename` should not be empty")
+                            .is_alphabetic();
+                        let is_identifier = first_char_alphabetic
+                            && str.chars().all(|c| c.is_alphanumeric() || c == '_');
+                        if !is_identifier {
+                            self.span_err(span, "Attribute `rename` should not be empty or should only contains alphanumeric characters and `_` or should start with a letter");
```
Comment (sonmarcho) [inline charon/src/translate/translate_ctx.rs:477]: "should not be empty": you already have a message for this above
"**or** should start": it is actually an **and**

File: charon/tests/ui/rename_attribute.rs [hunk 4]
```diff
@@ -37,11 +37,16 @@ enum SimpleEnum {
...
+#[charon::rename(StructTest)]
+struct Foo {
+    #[charon::rename(FieldTest)]
+    field1: u32,
+}
 
-#[aeneas::rename(TypeAeneas)]
-type Test2 = u32;
+#[charon::rename(Const_Test)]
+const SIX_SIX_SIX: u32 = 600 + 60 + 6;
```
Comment (sonmarcho) [inline charon/tests/ui/rename_attribute.rs:47]:
```suggestion
const C: u32 = 100 + 10 + 1;
```

File: charon/tests/ui/unsupported/rename_attribute_failure.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+//@ known-failure
+//@ no-check-output
```
Comment (sonmarcho) [inline charon/tests/ui/unsupported/rename_attribute_failure.rs:2]: We actually want to check the output, to make sure the error messages are correctly maintained

**Review by sonmarcho — CHANGES_REQUESTED** [review]

**Comment by sonmarcho** [comment]
Put aside the cleanup above, I would like to replace the `Span` field of `Field` with `ItemMeta` so that we can collect the attributes given to structure fields (and potentially rename fields).

**Comment by Nadrieril** [comment]
hm, if we use `ItemMeta` for things that aren't items then we shouldn't call it that

**Comment by R1kM** [comment]
I'd be in favour of wrapping up and merging this, and discussing the further extension to fields (and the precise name) in a further issue/PR. Wdyt?

File: charon/src/translate/translate_ctx.rs [hunk 6]
```diff
@@ -451,12 +451,55 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                            .chars()
+                            .nth(0)
+                            .expect("Attribute `rename` should not be empty")
+                            .is_alphabetic();
+                        let is_identifier = first_char_alphabetic
+                            && str
+                                .chars()
+                                .all(|c| c.is_alphanumeric() || c == '_' || c == '-');
+                        if !is_identifier {
+                            self.span_err(span, "Attribute `rename` should only contains alphanumeric characters and `_` or `-` and should start with a letter");
```
Comment (sonmarcho) [inline charon/src/translate/translate_ctx.rs:476]:
```suggestion
                            self.span_err(span, "Attribute `rename` should only contains alphanumeric character, `_` and `-`, and should start with a letter");
```

File: charon/src/translate/translate_types.rs [hunk 7]
```diff
@@ -601,12 +603,9 @@ impl<'tcx, 'ctx, 'ctx1> BodyTransCtx<'tcx, 'ctx, 'ctx1> {
                     }
                 };
 
-                // Translate the span information
-                let span = self.translate_span_from_rspan(field_def.span);
-
                 // Store the field
                 let field = Field {
-                    span,
+                    item_meta: item_meta.clone(),
```
Comment (sonmarcho) [inline charon/src/translate/translate_types.rs:608]: You're using the meta information of the *type* here, but you should use the type information from the *field*, which is different.
Ex.:
```rust
#[rename("Type")]
struct S {
  #[rename("y")]
  x : u32
}
```

**Review by sonmarcho — CHANGES_REQUESTED** [review]

File: charon/src/translate/translate_ctx.rs [hunk 8]
```diff
@@ -441,12 +441,55 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
         let opaque = attributes
             .iter()
             .any(|attr| attr == "charon::opaque" || attr == "aeneas::opaque");
+        let rename = {
+            let str = attributes.iter().find(|str| {
+                str.starts_with("charon::rename(") || str.starts_with("aeneas::rename(")
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:446]: `starts_with` is redundant with the `strip_prefix` that follows

Comment (Nadrieril) [reply charon/src/translate/translate_ctx.rs:446]: To make this code more robust, it would also be nice to have a `parse_attribute(&str) -> Option<Attribute>` function, with
```rust
enum Attribute {
  Opaque,
  Rename(String),
}
```

File: charon/src/ast/types.rs [hunk 9]
```diff
@@ -370,7 +370,7 @@ pub struct Variant {
 
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct Field {
-    pub span: Span,
+    pub item_meta: ItemMeta,
```
Comment (Nadrieril) [inline charon/src/ast/types.rs:373]: Please add a comment like `FIXME: define a more appropriate container for attribute and visibility information.`

File: charon/src/ast/meta.rs [hunk 10]
```diff
@@ -135,6 +135,7 @@ pub struct ItemMeta {
     /// API (this is called "pub-in-priv" items). With or without the `pub use`, we set `public =
     /// true`; computing item reachability is harder.
     pub public: bool,
+    pub rename: Option<String>,
```
Comment (Nadrieril) [inline charon/src/ast/meta.rs:138]: Please add a comment explaining what this field means

File: charon/src/translate/translate_types.rs [hunk 11]
```diff
@@ -584,7 +582,9 @@ impl<'tcx, 'ctx, 'ctx1> BodyTransCtx<'tcx, 'ctx, 'ctx1> {
             for (j, field_def) in var_def.fields.into_iter().enumerate() {
                 trace!("variant {i}: field {j}: {field_def:?}");
                 let field_span = field_def.span.rust_span_data.unwrap().span();
-
+                let item_meta = self
```
Comment (sonmarcho) [inline charon/src/translate/translate_types.rs:585]:
```suggestion
                let field_meta = self
```

File: charon/src/translate/translate_types.rs [hunk 12]
```diff
@@ -605,7 +605,7 @@ impl<'tcx, 'ctx, 'ctx1> BodyTransCtx<'tcx, 'ctx, 'ctx1> {
 
                 // Store the field
                 let field = Field {
-                    item_meta: item_meta.clone(),
+                    item_meta,
```
Comment (sonmarcho) [inline charon/src/translate/translate_types.rs:608]:
```suggestion
                    item_meta : field_meta,
```

**Review by sonmarcho — CHANGES_REQUESTED** [review]

**Comment by sonmarcho** [comment]
I just realized that we should also add some `ItemMeta` data to enumeration variants.

**Comment by Nadrieril** [comment]
Beware: another PR updated the `Cargo.toml` version, so when you merge `main` you'll have to update it again

File: charon/src/translate/translate_ctx.rs [hunk 13]
```diff
@@ -437,12 +437,64 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                                .chars()
+                                .all(|c| c.is_alphanumeric() || c == '_' || c == '-');
+                        if !is_identifier {
+                            self.span_err(span, "Attribute `rename` should only contains alphanumeric characters, `_` and `-` and should start with a letter");
+                            None
+                        } else {
+                            Some(rename.unwrap().to_string())
+                        }
+                    } else {
+                        self.span_err(span, "Attribute `rename` should not be empty");
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:471]: Don't use `span_err` directly, we have macros that do the right thing and register errors. Probably you want to return a `Result<.., Error>` and use the macro `register_error_or_panic`

Comment (Nadrieril) [reply charon/src/translate/translate_ctx.rs:471]: actually no, it would be silly to have to skip the whole item if an attribute is malformed. returning `ItemMeta` is fine, bug we need to register the error so charon knows there was one

Comment (Nadrieril) [reply charon/src/translate/translate_ctx.rs:471]: oh, `span_err` does that already, nevermind

File: charon/src/translate/translate_ctx.rs [hunk 14]
```diff
@@ -437,12 +437,64 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+            } else if filter_attribute.len() > 1 {
+                self.span_err(
+                    span,
+                    "There are too many `rename` attributes, please use only one",
+                );
+                None
+            } else {
+                None
+            }
+        };
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:490]: This whole thing is getting complex and hard to follow. Could you move it to another function, with early `return None` instead of nested `if`s? And please add some comments to explain the different phases of what you're doing.

File: charon/tests/crate_data.rs [hunk 15]
```diff
@@ -189,3 +189,127 @@ fn discriminants() -> Result<(), Box<dyn Error>> {
...
+
+        #[aeneas::rename("Type-Aeneas36")]
+        type Test2 = u32;
+        "#,
+    )?;
+
+    assert_eq!(
+        crate_data.trait_decls[0].item_meta.rename,
+        Some("BoolTest".to_string())
+    );
```
Comment (Nadrieril) [inline charon/tests/crate_data.rs:254]:
```suggestion
    assert_eq!(
        crate_data.trait_decls[0].item_meta.rename.as_deref(),
        Some("BoolTest")
    );
```

File: charon/tests/crate_data.rs [hunk 16]
```diff
@@ -189,3 +189,127 @@ fn discriminants() -> Result<(), Box<dyn Error>> {
...
+        Some("Const_Test".to_string())
+    );
+
+    assert_eq!(
+        crate_data.types[3].item_meta.rename,
+        Some("Type-Aeneas36".to_string())
+    );
+
+    let fieldvec = crate_data.types[2].kind.as_struct();
+    let field = fieldvec.get(0.into());
```
Comment (Nadrieril) [inline charon/tests/crate_data.rs:309]:
```suggestion
    let field = fieldvec[0];
```

File: charon/tests/crate_data.rs [hunk 17]
```diff
@@ -189,3 +189,127 @@ fn discriminants() -> Result<(), Box<dyn Error>> {
...
+    assert_eq!(
+        field.unwrap().item_meta.rename,
+        Some("FieldTest".to_string())
+    );
```
Comment (Nadrieril) [inline charon/tests/crate_data.rs:313]:
```suggestion
    let field = &crate_data.types[2].kind.as_struct()[0];
    assert_eq!(
        field.item_meta.rename.as_deref(),
        Some("FieldTest")
    );
```

File: charon/tests/ui/unsupported/rename_attribute_failure.rs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+//@ known-failure
```
Comment (Nadrieril) [inline charon/tests/ui/unsupported/rename_attribute_failure.rs:1]: The `unsupported` folder is for things we don't support yet but want to support eventually. Could you move this test out of the folder?

File: charon/src/translate/translate_ctx.rs [hunk 19]
```diff
@@ -426,6 +426,62 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                        .expect("Attribute `rename` should not be empty")
+                        .is_alphabetic();
+                    let is_identifier = first_char_alphabetic
+                        && str
+                            .chars()
+                            .all(|c| c.is_alphanumeric() || c == '_' || c == '-');
+                    if !is_identifier {
+                        self.span_err(span, "Attribute `rename` should only contains alphanumeric characters, `_` and `-` and should start with a letter");
+                        return None;
+                    } else {
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:461]: The point of asking you to use `return None;` is to avoid these nested `else` cases. The intent is that we can follow the code linearly where errors exit early and success continues down the flow of the function.

File: charon/src/translate/translate_ctx.rs [hunk 20]
```diff
@@ -399,8 +399,8 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
     fn register_file(&mut self, filename: FileName) -> FileId {
         // Lookup the file if it was already registered
         match self.translated.file_to_id.get(&filename) {
-            Some(id) => *id,
-            None => {
+            Option::Some(id) => *id,
+            Option::None => {
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:403]: Please don't re-add these `Option::`; they were removed in a recent PR

File: charon/src/translate/translate_ctx.rs [hunk 21]
```diff
@@ -426,6 +426,59 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                );
+                return None;
+            }
+        // if there are more than one rename attribute, we tell the user there should only be one
+        } else if filter_attribute.len() > 1 {
+            self.span_err(
+                span,
+                "There are too many `rename` attributes, please use only one",
+            );
+            return None;
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:476]: This could be an early return

File: charon/src/translate/translate_ctx.rs [hunk 22]
```diff
@@ -426,6 +426,59 @@ impl<'tcx, 'ctx> TranslateCtx<'tcx, 'ctx> {
...
+                    self.span_err(span, "Attribute `rename` should only contains alphanumeric characters, `_` and `-` and should start with a letter");
+                    return None;
+                }
+                return Some(rename.unwrap().to_string());
+            } else {
+                self.span_err(
+                    span,
+                    "Attribute `rename` should be of the shape `rename(\"...\")`",
+                );
+                return None;
```
Comment (Nadrieril) [inline charon/src/translate/translate_ctx.rs:468]: This could be an early return

**Review by sonmarcho — APPROVED** [review]
